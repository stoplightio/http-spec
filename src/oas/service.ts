import { isPlainObject } from '@stoplight/json';
import { DeepPartial, IHttpService } from '@stoplight/types';
import pickBy = require('lodash.pickby');
import { OpenAPIObject } from 'openapi3-ts';
import { Spec } from 'swagger-schema-official';

import { isBoolean, isNonNullable, isString } from '../guards';
import { TranslateFunction } from '../types';
import { hasXLogo } from './guards';
import { translateTagDefinition } from './tags';
import { translateLogo } from './transformers/translateLogo';

export const transformOasService: TranslateFunction<DeepPartial<OpenAPIObject> | DeepPartial<Spec>, [], IHttpService> =
  function () {
    const document = this.document;
    const id = String(document['x-stoplight']?.id);
    this.ids.service = id;
    this.parentId = id;

    const httpService: IHttpService = {
      id,

      version: document.info?.version ?? '',
      name: document.info?.title ?? 'no-title',

      ...pickBy(
        {
          description: document.info?.description,
          termsOfService: document.info?.termsOfService,
        },
        isString,
      ),

      ...pickBy(
        {
          contact: document.info?.contact,
        },
        isPlainObject,
      ),

      ...pickBy(
        {
          internal: document['x-internal'],
        },
        isBoolean,
      ),
    };

    if (isPlainObject(document.info) && hasXLogo(document.info)) {
      httpService.logo = translateLogo(document.info);
    }

    const tags = Array.isArray(document.tags)
      ? document.tags.map(translateTagDefinition, this).filter(isNonNullable)
      : [];

    if (tags.length > 0) {
      httpService.tags = tags;
    }

    return httpService;
  };
