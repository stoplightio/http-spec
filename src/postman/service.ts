import { Collection, CollectionDefinition, Item, ItemGroup } from 'postman-collection';
import { HttpServiceTransformer } from '../types';
import {
  isSecuritySchemeOutcomeEqual,
  isStandardSecurityScheme,
  SecuritySchemeOutcome,
  transformSecurityScheme,
} from './transformers/securityScheme';
import { traverseItemsAndGroups } from './util';

export const transformPostmanCollectionService: HttpServiceTransformer<CollectionDefinition> = collectionDefinition => {
  const collection = new Collection(collectionDefinition);
  const resolvedCollection = new Collection(collection.toObjectResolved({ variables: collection.variables }, []));

  const securitySchemeOutcomes: SecuritySchemeOutcome[] = [];
  let securitySchemeIdx = 0;

  function addSecurityScheme(outcome: SecuritySchemeOutcome) {
    if (!securitySchemeOutcomes.find(o => isSecuritySchemeOutcomeEqual(o, outcome))) {
      securitySchemeOutcomes.push(outcome);
    }
  }

  traverseItemsAndGroups(
    (resolvedCollection as unknown) as ItemGroup<Item>,
    item => {
      const auth = item.getAuth();
      if (auth) {
        const transformed = transformSecurityScheme(auth, type => `${type}-${securitySchemeIdx++}`);
        if (transformed) addSecurityScheme(transformed);
      }
    },
    itemGroup => {
      if (itemGroup.auth) {
        const transformed = transformSecurityScheme(itemGroup.auth, type => `${type}-${securitySchemeIdx++}`);
        if (transformed) addSecurityScheme(transformed);
      }
    },
  );

  return {
    id: resolvedCollection.id,
    name: resolvedCollection.name,
    version: resolvedCollection.version
      ? resolvedCollection.version.string
        ? resolvedCollection.version.string
        : `${resolvedCollection.version.major}.${resolvedCollection.version.minor}.${resolvedCollection.version.patch}-${resolvedCollection.version.prerelease}`
      : '1.0.0',
    description: resolvedCollection.description?.toString(),
    securitySchemes: securitySchemeOutcomes.filter(isStandardSecurityScheme).map(outcome => outcome.securityScheme),
  };
};
