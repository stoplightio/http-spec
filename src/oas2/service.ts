import { isPlainObject } from '@stoplight/json';
import { HttpSecurityScheme, IHttpOperation, IHttpOperationResponse, Optional, Reference } from '@stoplight/types';
import pickBy = require('lodash.pickby');

import { withContext } from '../context';
import { isNonNullable, isString } from '../guards';
import { createContext } from '../oas/context';
import { bundleResolveRef } from '../oas/resolver';
import { transformOasService } from '../oas/service';
import { translateToComponents } from '../oas/transformers/components';
import { translateSchemaObjectFromPair } from '../oas/transformers/schema';
import { Oas2HttpServiceBundle, Oas2HttpServiceTransformer, OasVersion } from '../oas/types';
import { ArrayCallbackParameters } from '../types';
import { entries } from '../utils';
import { normalizeProducesOrConsumes } from './accessors';
import { transformOas2Operations } from './operation';
import { translateToSharedParameters } from './transformers/params';
import { translateToResponse } from './transformers/responses';
import { translateToSingleSecurity } from './transformers/securities';
import { translateToServer } from './transformers/servers';
import { Oas2TranslateFunction } from './types';

export const bundleOas2Service: Oas2HttpServiceBundle = ({ document: _document }) => {
  const ctx = createContext(_document, bundleResolveRef);
  const { document } = ctx;

  const { securitySchemes, ...service } = transformOas2Service({ document, ctx });
  const produces = normalizeProducesOrConsumes(ctx.document.produces);
  const components = {
    ...translateToComponents.call(ctx, OasVersion.OAS2, {
      responses([key, value]): Optional<
        IHttpOperationResponse<true> | (Pick<IHttpOperationResponse<true>, 'code'> & Reference)
      > {
        return translateToResponse.call(ctx, produces, key, value);
      },
      definitions: translateSchemaObjectFromPair,
      securitySchemes: translateSecurityScheme,
    }),
    ...translateToSharedParameters.call(ctx, document),
  } as any;

  const operations = transformOas2Operations(document, ctx) as unknown as IHttpOperation<true>[];

  return {
    ...service,
    operations,
    components,
  };
};

export const transformOas2Service: Oas2HttpServiceTransformer = ({ document, ctx = createContext(document) }) => {
  const httpService = transformOasService.call(ctx);

  if (document.info?.license) {
    httpService.license = {
      ...document.info.license,
      name: document.info.license.name || '',
    };
  }

  const schemes = Array.isArray(document.schemes) ? document.schemes.filter(isString) : [];

  const servers = schemes.map(translateToServer, ctx).filter(isNonNullable);

  if (servers.length) {
    httpService.servers = servers;
  }

  const securitySchemes = entries(document.securityDefinitions).map(translateSecurityScheme, ctx).filter(isNonNullable);

  if (securitySchemes.length) {
    httpService.securitySchemes = securitySchemes;
  }

  const security = Array.isArray(document.security)
    ? document.security
        .flatMap(sec => {
          if (!isPlainObject(sec)) return null;

          return Object.keys(sec).map(key => {
            const ss = securitySchemes.find(securityScheme => securityScheme.key === key);
            if (ss && ss.type === 'oauth2') {
              const flows = {};
              for (const flowKey in ss.flows) {
                const flow = ss.flows[flowKey];
                flows[flowKey] = {
                  ...flow,
                  scopes: pickBy(flow.scopes, (_val: string, scopeKey: string) => {
                    const secKey = sec[key];
                    if (secKey) return secKey.includes(scopeKey);
                    return undefined;
                  }),
                };
              }

              return {
                ...ss,
                flows,
              };
            }

            return ss;
          });
        })
        .filter(isNonNullable)
    : [];

  if (security.length) {
    httpService.security = security;
  }

  return httpService;
};

const translateSecurityScheme = withContext<
  Oas2TranslateFunction<ArrayCallbackParameters<[name: string, scheme: unknown]>, Optional<HttpSecurityScheme>>
>(function ([key, definition]) {
  if (!isPlainObject(definition)) return;

  return translateToSingleSecurity.call(this, { ...definition, key });
});
