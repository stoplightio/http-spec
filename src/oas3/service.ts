import { HttpSecurityScheme, IHttpService, INodeTag, IServer } from '@stoplight/types';
import { compact, filter, flatMap, keys, map } from 'lodash';

import { Oas3HttpServiceTransformer } from '../oas/types';
import { isSecurityScheme, isTagObject } from './guards';
import { transformToSingleSecurity } from './transformers/securities';
import { translateServerVariables } from './transformers/servers';

export const transformOas3Service: Oas3HttpServiceTransformer = ({ document }) => {
  const httpService: IHttpService = {
    id: '?http-service-id?',
    version: document.info?.version ?? '',
    name: document.info?.title ?? 'no-title',
  };

  if (document.info?.description) {
    httpService.description = document.info.description;
  }

  if (document.info?.contact) {
    httpService.contact = document.info.contact;
  }

  if (document.info?.license) {
    httpService.license = {
      ...document.info.license,
      name: document.info.license.name || '',
    };
  }

  if (document.info?.termsOfService) {
    httpService.termsOfService = document.info.termsOfService;
  }

  const servers = compact<IServer>(
    map(document.servers, server => {
      if (!server) return null;

      const serv: IServer = {
        name: document.info?.title ?? '',
        description: server.description,
        url: server.url ?? '',
      };

      const variables = server.variables && translateServerVariables(server.variables);
      if (variables && Object.keys(variables).length) serv.variables = variables;

      return serv;
    }),
  );
  if (servers.length) {
    httpService.servers = servers;
  }

  const securitySchemes = compact<HttpSecurityScheme>(
    keys(document.components?.securitySchemes).map(key => {
      const definition = document?.components?.securitySchemes?.[key];
      return isSecurityScheme(definition) && transformToSingleSecurity(definition, key);
    }),
  );
  if (securitySchemes.length) {
    httpService.securitySchemes = securitySchemes;
  }

  const security = compact(
    flatMap(document.security, sec => {
      if (!sec) return null;

      return keys(sec).map(key => {
        return securitySchemes.find(securityScheme => {
          return securityScheme.key === key;
        });
      });
    }),
  );
  if (security.length) {
    httpService.security = security;
  }

  const tags = filter(document.tags, isTagObject);
  if (tags.length) {
    httpService.tags = tags;
  }

  return httpService;
};
