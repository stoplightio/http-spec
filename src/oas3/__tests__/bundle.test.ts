import { bundleOas3Service } from '../service';

test('works', () => {
  expect(
    bundleOas3Service({
      document: {
        openapi: '3.1.0',
        'x-stoplight': {
          id: 'service_abc',
        },
        info: {
          title: 'Users API',
          version: '1.0',
        },
        servers: [
          {
            url: 'http://localhost:3000',
          },
        ],
        tags: [
          {
            name: 'mutates',
          },
        ],
        paths: {
          '/users/{userId}': {
            parameters: [
              {
                schema: {
                  type: 'integer',
                },
                name: 'userId',
                in: 'path',
                required: true,
                description: 'Id of an existing user.',
              },
              {
                $ref: '#/components/parameters/Some-Header',
              },
            ],
            get: {
              operationId: 'get-user',
              summary: 'Get User Info by User ID',
              description: 'Retrieve the information of the user with the matching user ID.',
              tags: ['tag-without-root-def'],
              parameters: [
                {
                  schema: {
                    type: 'boolean',
                  },
                  in: 'query',
                  name: 'summaryOnly',
                },
              ],
              responses: {
                '200': {
                  description: 'User Found',
                  content: {
                    'application/json': {
                      schema: {
                        $ref: '#/components/schemas/User',
                      },
                    },
                    'application/xml': {
                      schema: {
                        type: 'string',
                      },
                    },
                  },
                },
                '404': {
                  $ref: '#/components/responses/ErrorResponse',
                },
              },
            },
            post: {
              operationId: 'post-users-userId',
              summary: 'Create user',
              tags: ['mutates'],
              parameters: [
                {
                  schema: {
                    type: 'integer',
                  },
                  name: 'Post-Specific-Header',
                  in: 'header',
                },
              ],
              security: [
                {
                  'api-key': [],
                },
              ],
              requestBody: {
                content: {
                  'application/json': {
                    schema: {
                      $ref: '#/components/schemas/User',
                    },
                    examples: {
                      'basic-example': {
                        $ref: '#/components/examples/A-Shared-Example',
                      },
                    },
                  },
                },
              },
              responses: {
                '201': {
                  description: 'User Created',
                  content: {
                    'application/json': {
                      schema: {
                        $ref: '#/components/schemas/User',
                      },
                    },
                  },
                },
                '400': {
                  $ref: '#/components/responses/ErrorResponse',
                },
              },
            },
          },
        },
        components: {
          schemas: {
            User: {
              title: 'User',
              type: 'object',
              properties: {
                id: {
                  type: 'integer',
                  readOnly: true,
                },
                address: {
                  $ref: '#/components/schemas/Address',
                },
              },
              required: ['id'],
              examples: [],
            },
            Address: {
              title: 'Address',
              type: 'object',
              properties: {
                street: {
                  type: 'string',
                },
              },
              examples: [
                {
                  street: '422 W Riverside Drive',
                },
              ],
            },
            UserId: {
              type: 'number',
              title: 'UserId',
              minimum: 0,
            },
            Error: {
              title: 'Error',
              type: 'object',
              properties: {
                code: {
                  type: 'number',
                },
                msg: {
                  type: 'string',
                },
              },
              required: ['code', 'msg'],
            },
          },
          parameters: {
            'Some-Header': {
              name: 'A-Shared-Header',
              in: 'header',
              required: false,
              schema: {
                type: 'string',
              },
            },
          },
          examples: {
            'A-Shared-Example': {
              value: {
                id: 0,
                address: {
                  street: 'string',
                },
              },
            },
          },
          securitySchemes: {
            'api-key': {
              name: 'API Key',
              type: 'apiKey',
              in: 'query',
            },
          },
          responses: {
            ErrorResponse: {
              description: 'A generic error response.',
              content: {
                'application/json': {
                  schema: {
                    type: 'object',
                    properties: {
                      error: {
                        $ref: '#/components/schemas/Error',
                      },
                    },
                  },
                },
              },
            },
          },
        },
      },
    }),
  ).toMatchSnapshot();
});
