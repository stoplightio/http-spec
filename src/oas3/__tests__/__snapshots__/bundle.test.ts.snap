// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works 1`] = `
Object {
  "components": Object {
    "examples": Object {
      "A-Shared-Example": Object {
        "id": "5a69041e065b0",
        "key": "A-Shared-Example",
        "value": Object {
          "address": Object {
            "street": "string",
          },
          "id": 0,
        },
      },
    },
    "parameters": Object {
      "Some-Header": Object {
        "examples": Array [],
        "id": "21b1f96bd26ee",
        "name": "A-Shared-Header",
        "required": false,
        "schema": Object {
          "$schema": "http://json-schema.org/draft-07/schema#",
          "type": "string",
          "x-stoplight": Object {
            "id": "be6b513de1b69",
          },
        },
        "style": "simple",
      },
    },
    "requestBodies": Object {},
    "responses": Object {
      "ErrorResponse": Object {
        "code": "ErrorResponse",
        "contents": Array [
          Object {
            "encodings": Array [],
            "examples": Array [],
            "id": "4d98be34f341a",
            "mediaType": "application/json",
            "schema": Object {
              "$schema": "http://json-schema.org/draft-07/schema#",
              "properties": Object {
                "error": Object {
                  "$ref": "#/components/schemas/Error",
                },
              },
              "type": "object",
              "x-stoplight": Object {
                "id": "2691eb0db9efc",
              },
            },
          },
        ],
        "description": "A generic error response.",
        "headers": Array [],
        "id": "437771f63f179",
      },
    },
    "schemas": Object {
      "Address": Object {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "examples": Array [
          Object {
            "street": "422 W Riverside Drive",
          },
        ],
        "properties": Object {
          "street": Object {
            "type": "string",
          },
        },
        "title": "Address",
        "type": "object",
        "x-stoplight": Object {
          "id": "8b1562dcd786e",
        },
      },
      "Error": Object {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "properties": Object {
          "code": Object {
            "type": "number",
          },
          "msg": Object {
            "type": "string",
          },
        },
        "required": Array [
          "code",
          "msg",
        ],
        "title": "Error",
        "type": "object",
        "x-stoplight": Object {
          "id": "da49423a5a9e7",
        },
      },
      "User": Object {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "examples": Array [],
        "properties": Object {
          "address": Object {
            "$ref": "#/components/schemas/Address",
          },
          "id": Object {
            "readOnly": true,
            "type": "integer",
          },
        },
        "required": Array [
          "id",
        ],
        "title": "User",
        "type": "object",
        "x-stoplight": Object {
          "id": "de4f083463b7c",
        },
      },
      "UserId": Object {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "minimum": 0,
        "title": "UserId",
        "type": "number",
        "x-stoplight": Object {
          "id": "a8978c71e4d00",
        },
      },
    },
    "securitySchemes": Object {
      "api-key": Object {
        "id": "202a905f9dff6",
        "in": "query",
        "key": "api-key",
        "name": "API Key",
        "type": "apiKey",
      },
    },
  },
  "id": "service_abc",
  "name": "Users API",
  "operations": Array [
    Object {
      "description": "Retrieve the information of the user with the matching user ID.",
      "extensions": Object {},
      "id": "96043a63b6901",
      "iid": "get-user",
      "method": "get",
      "path": "/users/{userId}",
      "request": Object {
        "body": Object {
          "contents": Array [],
          "id": "d5027559477f8",
        },
        "cookie": Array [],
        "headers": Array [
          Object {
            "$ref": "#/components/parameters/Some-Header",
          },
        ],
        "path": Array [
          Object {
            "description": "Id of an existing user.",
            "examples": Array [],
            "id": "fe171ec8cfd0b",
            "name": "userId",
            "required": true,
            "schema": Object {
              "$schema": "http://json-schema.org/draft-07/schema#",
              "type": "integer",
              "x-stoplight": Object {
                "id": "13ad531bed72e",
              },
            },
            "style": "simple",
          },
        ],
        "query": Array [
          Object {
            "examples": Array [],
            "id": "efe9534d001fc",
            "name": "summaryOnly",
            "schema": Object {
              "$schema": "http://json-schema.org/draft-07/schema#",
              "type": "boolean",
              "x-stoplight": Object {
                "id": "aca62504578bd",
              },
            },
            "style": "form",
          },
        ],
      },
      "responses": Array [
        Object {
          "code": "200",
          "contents": Array [
            Object {
              "encodings": Array [],
              "examples": Array [],
              "id": "fce50f391bf57",
              "mediaType": "application/json",
              "schema": Object {
                "$ref": "#/components/schemas/User",
              },
            },
            Object {
              "encodings": Array [],
              "examples": Array [],
              "id": "48eeb3ee2a049",
              "mediaType": "application/xml",
              "schema": Object {
                "$schema": "http://json-schema.org/draft-07/schema#",
                "type": "string",
                "x-stoplight": Object {
                  "id": "069dfbb6c6315",
                },
              },
            },
          ],
          "description": "User Found",
          "headers": Array [],
          "id": "f387e16c7d39d",
        },
        Object {
          "$ref": "#/components/responses/ErrorResponse",
        },
      ],
      "security": Array [],
      "servers": Array [
        Object {
          "id": "98ddf8a4b5bdc",
          "name": "Users API",
          "url": "http://localhost:3000",
        },
      ],
      "summary": "Get User Info by User ID",
      "tags": Array [
        Object {
          "id": "9862017e672e6",
          "name": "tag-without-root-def",
        },
      ],
    },
    Object {
      "extensions": Object {},
      "id": "b16a96d287951",
      "iid": "post-users-userId",
      "method": "post",
      "path": "/users/{userId}",
      "request": Object {
        "body": Object {
          "contents": Array [
            Object {
              "encodings": Array [],
              "examples": Array [
                Object {
                  "$ref": "#/components/examples/A-Shared-Example",
                },
              ],
              "id": "00db77c676e1a",
              "mediaType": "application/json",
              "schema": Object {
                "$ref": "#/components/schemas/User",
              },
            },
          ],
          "id": "07267ec331fc9",
        },
        "cookie": Array [],
        "headers": Array [
          Object {
            "examples": Array [],
            "id": "1ead595922478",
            "name": "Post-Specific-Header",
            "schema": Object {
              "$schema": "http://json-schema.org/draft-07/schema#",
              "type": "integer",
              "x-stoplight": Object {
                "id": "de5897f178a5d",
              },
            },
            "style": "simple",
          },
          Object {
            "$ref": "#/components/parameters/Some-Header",
          },
        ],
        "path": Array [
          Object {
            "description": "Id of an existing user.",
            "examples": Array [],
            "id": "fe171ec8cfd0b",
            "name": "userId",
            "required": true,
            "schema": Object {
              "$schema": "http://json-schema.org/draft-07/schema#",
              "type": "integer",
              "x-stoplight": Object {
                "id": "13ad531bed72e",
              },
            },
            "style": "simple",
          },
        ],
        "query": Array [],
      },
      "responses": Array [
        Object {
          "code": "201",
          "contents": Array [
            Object {
              "encodings": Array [],
              "examples": Array [],
              "id": "88460a8f1a612",
              "mediaType": "application/json",
              "schema": Object {
                "$ref": "#/components/schemas/User",
              },
            },
          ],
          "description": "User Created",
          "headers": Array [],
          "id": "d8ca38606ee5d",
        },
        Object {
          "$ref": "#/components/responses/ErrorResponse",
        },
      ],
      "security": Array [
        Array [
          Object {
            "id": "202a905f9dff6",
            "in": "query",
            "key": "api-key",
            "name": "API Key",
            "type": "apiKey",
          },
        ],
      ],
      "servers": Array [
        Object {
          "id": "98ddf8a4b5bdc",
          "name": "Users API",
          "url": "http://localhost:3000",
        },
      ],
      "summary": "Create user",
      "tags": Array [
        Object {
          "id": "936737e88c6fa",
          "name": "mutates",
        },
      ],
    },
  ],
  "servers": Array [
    Object {
      "id": "98ddf8a4b5bdc",
      "name": "Users API",
      "url": "http://localhost:3000",
    },
  ],
  "tags": Array [
    Object {
      "id": "936737e88c6fa",
      "name": "mutates",
    },
  ],
  "version": "1.0",
}
`;
